### ISSUE TITLE
Test clusters regularly fail health checks - kube-dns cycles through Ready/NotReady

### ISSUE THREAD

**Issue by quinton-hoole** [GoogleCloudPlatform/kubernetes#10507]
... and as a result, e2e tests are not run.

One common failure mode is that some of our static system containers never report ready.
```
INFO: 4 / 6 pods in namespace 'default' are running and ready (390 seconds elapsed)
INFO: POD                           NODE                 PHASE    CONDITIONS
INFO: kube-dns-v4-jtto1             e2e-gce-minion-2d59  Running  [{Ready False}]
INFO: monitoring-heapster-v5-1fwj5  e2e-gce-minion-eeif  Running  [{Ready False}]
INFO:
```
If you zoom out a bit, you'll notice that at least the kube-dns pod oscillates between Ready=True and Ready=False.

e.g. 

```
[15] INFO: 4 / 6 pods in namespace 'default' are running and ready (550 seconds elapsed)
[15] INFO: POD                           NODE                    PHASE    CONDITIONS
[15] INFO: kube-dns-v4-21fk7             pull-e2e-2-minion-y29d  Running  [{Ready False}]
[15] INFO: monitoring-heapster-v5-9z944  pull-e2e-2-minion-y29d  Running  [{Ready False}]
[15] INFO: 
[15] INFO: 4 / 6 pods in namespace 'default' are running and ready (555 seconds elapsed)
[15] INFO: POD                           NODE                    PHASE    CONDITIONS
[15] INFO: kube-dns-v4-21fk7             pull-e2e-2-minion-y29d  Running  [{Ready False}]
[15] INFO: monitoring-heapster-v5-9z944  pull-e2e-2-minion-y29d  Running  [{Ready False}]
[15] INFO: 
[15] INFO: 5 / 6 pods in namespace 'default' are running and ready (560 seconds elapsed)
[15] INFO: POD                           NODE                    PHASE    CONDITIONS
[15] INFO: monitoring-heapster-v5-9z944  pull-e2e-2-minion-y29d  Running  [{Ready False}]
[15] INFO: 
[15] INFO: 5 / 6 pods in namespace 'default' are running and ready (565 seconds elapsed)
[15] INFO: POD                           NODE                    PHASE    CONDITIONS
[15] INFO: monitoring-heapster-v5-9z944  pull-e2e-2-minion-y29d  Running  [{Ready False}]
[15] INFO: 
[15] INFO: 5 / 6 pods in namespace 'default' are running and ready (570 seconds elapsed)
[15] INFO: POD                           NODE                    PHASE    CONDITIONS
[15] INFO: monitoring-heapster-v5-9z944  pull-e2e-2-minion-y29d  Running  [{Ready False}]
[15] INFO: 
[15] INFO: 5 / 6 pods in namespace 'default' are running and ready (575 seconds elapsed)
[15] INFO: POD                           NODE                    PHASE    CONDITIONS
[15] INFO: monitoring-heapster-v5-9z944  pull-e2e-2-minion-y29d  Running  [{Ready False}]
[15] INFO: 
[15] INFO: 4 / 6 pods in namespace 'default' are running and ready (580 seconds elapsed)
[15] INFO: POD                           NODE                    PHASE    CONDITIONS
[15] INFO: kube-dns-v4-21fk7             pull-e2e-2-minion-y29d  Running  [{Ready False}]
[15] INFO: monitoring-heapster-v5-9z944  pull-e2e-2-minion-y29d  Running  [{Ready False}]
[15] INFO: 
[15] INFO: 4 / 6 pods in namespace 'default' are running and ready (585 seconds elapsed)
[15] INFO: POD                           NODE                    PHASE    CONDITIONS
[15] INFO: kube-dns-v4-21fk7             pull-e2e-2-minion-y29d  Running  [{Ready False}]
[15] INFO: monitoring-heapster-v5-9z944  pull-e2e-2-minion-y29d  Running  [{Ready False}]
[15] INFO: 
```

**Comment by lavalamp** [comment]
This must have broken very recently. @vishh did a new heapster get pushed? @thockin did a new kube2sky get pushed?

**Comment by quinton-hoole** [comment]
It appears to have started around Jenkins job/kubernetes-e2e-gce/7296/ at ~ 10:30 AM today.

**Comment by lavalamp** [comment]
I really don't see anything suspicious that merged at that time.

**Comment by quinton-hoole** [comment]
Note that the failure is intermittent.  It's happening about 50% of the time.

It's entirely possible that kube-dns has been flaky like this for a while, and now that monitoring-heapster is not becoming ready at all, that's necessary and sufficient to cause the cluster validation to fail.

**Comment by a-robinson** [comment]
kube-dns was updated on Friday by #10150

**Comment by quinton-hoole** [comment]
OK, that correlates roughly and anecdotally with when e2e tests became more flaky in general.
I suspect that we have a more recent breakage (this morning) causing monitoring-heapster to intermittently not become Ready at all.

Action items: 
1. Gather some e2e test failure data to corroborate my claim above about kube-dns update on Friday being correlated with increased e2e test failure rates.
2. Analyse cluster logs to figure out why exactly monitoring-heapster is sometimes not becoming Ready.

@lavalamp want to take a stab at 2?  I can do 1, which shouldn't take very long.

**Comment by quinton-hoole** [comment]
Regarding 1, the DNS e2e test specifically got flaky a few days ago.  Busy getting a more accurate timeline, but I think that's a smoking gun.

**Comment by quinton-hoole** [comment]
Jenkins build 7242 on Sat Jun 27, 2015 4:02 AM was the first DNS e2e test failure in a long while.  Some cluster builds in the hours before that are also potentially attributable to DNS failures (which prevented the tests from running at all). 

The DNS change in #10150 merged at 6:17 PM on Fri June 26, i.e. 10 hours before the first official DNS test failure.

**Comment by quinton-hoole** [comment]
Gotta go - back in a bit.

**Comment by bprashanth** [comment]
Maybe we can have an e2e that checks for the cluster addons, and continue testing everything else even when they're down? I don't see why we need to declare the cluster dead

**Comment by quinton-hoole** [comment]
Well, when for example DNS is dead, a whole bunch of other tests tend to fail, understandably.  And also, several tests are interfered with by cluster addons busy scheduling or rescheduling while the test is running.  I think we should just fix the addons :-)

**Comment by bprashanth** [comment]
The same applies for any critical component (eg: rcs, scheduler). If you have an isolated test, devs can debug without holding up traffic.

**Comment by mikedanese** [comment]
heapster was version bumped today :( #10369

**Comment by dchen1107** [comment]
I just filed #10513 for skydns and #10511 for heapster

**Comment by lavalamp** [comment]
Ah. In that case, I won't bother debugging it. @mikedanese I think we should roll that back. :(

**Comment by a-robinson** [comment]
I believe the root cause of both problems is the same -- the service account they expect (/var/run/secrets/kubernetes.io/serviceaccount/token) isn't there

**Comment by lavalamp** [comment]
#10514

**Comment by mikedanese** [comment]
yup. there is an e2e test for that too. https://github.com/GoogleCloudPlatform/kubernetes/blob/master/test/e2e/service_accounts.go. cc @liggitt as he is the serviceaccount expert. what type of latency can we expect on mounting service account secrets?

**Comment by dchen1107** [comment]
I can see #10369 is obviously contributing to heapster failure (#10511). But I am not sure about skydns case here.

**Comment by quinton-hoole** [comment]
Yeas @mikedanese , that merge of #10369 at 10:56 AM today correlates precisely with the beginning of the heapster problem, which is the cause of the failed cluster builds.  Lets fix that in dawn's #10511

**Comment by liggitt** [comment]
Latency only applies to service account tokens on initial namespace creation. I don't have the relevant tests in front of me… are they creating a new namespace or running tests in an existing namespace? 

When the shared controller manager client isn't under load, I typically see the tokens get provisioned in < 100ms from namespace creation time. Under load, I've seen as high as 5-10 seconds (typically means the controller manager client's QPS is exhausted)

**Comment by lavalamp** [comment]
It seems like the actual problem is indeed the service account secrets not being added to the pod.

**Comment by liggitt** [comment]
Also, as we started relying more on service account tokens in openshift, we discovered several environments that had issues with the underlying tmpfs secret mount mechanism… not sure if any of those are in play here. @pmorie, do you know if we have any relevant fixes still in flight?

**Comment by lavalamp** [comment]
@liggitt it's the default namespace; the addon pods are coming up without service account tokens some of the time.

**Comment by liggitt** [comment]
do the pod specs in the API have the service account field set and no corresponding secret volume mount?

**Comment by dchen1107** [comment]
Ok, we should close that #10369 since #10511 is not the root case. In good case, the kubectl get pods -o yaml should have volumeMount as the below, but the failure cases don't have those blob:

```
    volumeMounts:
    - mountPath: /var/run/secrets/kubernetes.io/serviceaccount
      name: default-token-4lyfe
      readOnly: true
```

**Comment by lavalamp** [comment]
@liggitt Yes, exactly.

**Comment by a-robinson** [comment]
Dumb question time - are these secrets being auto-populated? If so, by what? The dns and heapster RC configs don't explicitly ask for them.

**Comment by lavalamp** [comment]
@a-robinson Yes, there's an admission plug-in that adds them IIRC.

**Comment by mikedanese** [comment]
@a-robinson token_controller and serviceaccount_controller mostly. We add the default serviceaccount to pods. The default service account has two secrets: a bearer token and the apiserver root ca.crt which are mounted in all pods. The implementation of pkg/client.InClusterConfig() relies on this.

**Comment by quinton-hoole** [comment]
Also, can anyone confirm whether the service account tokens never arrive vs arrive after they're referenced by e.g. sky-dns?  Perhaps pods requiring service account tokens need to retry?

**Comment by liggitt** [comment]
Yes, all pods get a default service account assigned, and a mounted API token for their specified service account, if one exists. The default service account and API token is autocreated when a namespace is created.

**Comment by liggitt** [comment]
Admission could reject pods if their service account doesn't have a token, but I'm doubtful everything would react well to that

**Comment by liggitt** [comment]
I would prefer pods have a way to signal they want/need a service account token, but that doesn't exist yet.

**Comment by thockin-cc** [comment]
We need some sort of interlock here - telling people to delete their DNS
pod to fix a problems is ... yeah, what;s the real answer?

**Comment by lavalamp** [comment]
@quinton-hoole they never arrive; deleting the pod (letting the RC make another) fixes the problem, though.

**Comment by liggitt** [comment]
Admission is the only chance to add the mount. Can't mutate post-create.

**Comment by lavalamp** [comment]
We could fix in the addons script by waiting for the default namespace to be ready, perhaps?

**Comment by quinton-hoole** [comment]
@ligget Why?  We hot mount other devices, why not these?

**Comment by quinton-hoole** [comment]
@lavalamp, that sounds very sensible.

**Comment by lavalamp** [comment]
I will make a fix then

**Comment by lavalamp** [comment]
@quinton-hoole it's part of the pod spec, which cannot be mutated for whatever reason. It's not a kubelet problem, it's master screwing up the pod definition.

**Comment by liggitt** [comment]
The name of the secret isn't known prior to create (it is generated) so the spec for the volume mount can't be built.

**Comment by liggitt** [comment]
Until initializers (or equivalent) exist, our options are limited to rejecting pods until preconditions exist or making clients wait for preconditions themselves. Not sure which I dislike more. Rejecting is "purer" but requires clients to handle retrying correctly.

**Comment by lavalamp** [comment]
I think rejecting with an error would be a much more graceful user experience. I agree that it would be a terrible hack to make admission actually wait.

**Comment by liggitt** [comment]
Rejecting with an error is easy but high impact, especially when we do so for pods that don't care about API tokens. I guess RC-controlled pods would handle that well enough, but things creating run-once pods likely wouldn't.

**Comment by lavalamp** [comment]
It's only an issue in the few seconds after a namespace is created, though, right?

**Comment by liggitt** [comment]
> I guess RC-controlled pods would handle that well enough

Though the fact that we had to do special things to make our e2e tests handle rejections when the default service account wasn't provisioned yet makes me suspect rejections over API tokens wouldn't be generally handled well either

**Comment by liggitt** [comment]
Yes, only immediately after a namespace creation

**Comment by lavalamp** [comment]
And if RCs drive your pods, everything should Just Work...

**Comment by thockin** [comment]
Why is it a hack to wait?

**Comment by liggitt** [comment]
> And if RCs drive your pods, everything should Just Work...

Yes

> Why is it a hack to wait?

Making a create pod call wait synchronously for something to be created from another asynchronous controller seems counter to the whole approach of the API

**Comment by lavalamp** [comment]
It's a hack to be doing it in admission at all, this is an initializer.

**Comment by liggitt** [comment]
#10523 enables admission rejecting pods until API tokens exist. There was already the possibility of rejection if the service account wasn't provisioned yet, so I'm leaning more toward this approach until we can convert to an initializer approach. RCs should keep retrying, so impact would be limited to other run-once pod-creators. As more components depend on service account tokens, that impact becomes more justified... I'll defer to @lavalamp, @thockin, or other interested parties on whether this is what we want to do for 1.0

**Comment by lavalamp** [comment]
#10520 should be good to go now. #10523 looks good, too. Possibly we want both, but we need one immediately to stop the bleeding. Someone else can decide, I have to go home.

**Comment by quinton-hoole** [comment]
I'll take a look at both, and try to merge one of them tonight.

**Comment by thockin** [comment]
I am OK with rejection

**Comment by derekwaynecarr** [comment]
I have started documenting an approach for initializers since we really need to do something post 1.0, I think for e2e it's fine to not wait in admission

**Comment by wojtek-t** [comment]
+1 for rejection from #10523

**Comment by quinton-hoole** [comment]
Fixed by #10523
